### PR TITLE
Avoid scroll bleed when displaying modal UI on mobile

### PR DESCRIPTION
**Pull request by brandonpayton**
## Description
This is a PR for avoiding scroll bleed when displaying modal UI on mobile devices. This is meant to address the first part of #4082.

## How Has This Been Tested?

I successfully ran the unit tests. I also tried running the e2e tests but had an inconsistent experience with different failures with each run.

I manually tested using an iPhone 6s simulator, an iPhone X simulator, and a Galaxy S5 with Chrome:

1. Observing that I could scroll the editor body
2. Opening block insertion modal and observing that I could only scroll the menu, not the body
3. Closing the block insertion modal and observing I could scroll the body
4. Opening the sidebar modal and observing that I could scroll the sidebar but not the body.
5. Closing the sidebar and observing that I could scroll the body
6. Opening the publish options modal, expanding sections so the modal could scroll, and observing that I could scroll the sidebar but not the body.
7. Closing the publish options modal and observing that I could scroll the body.

## Types of changes

This PR updates `Popover` to prevent body scrolling while a modal popover is open. It also adds a `Popover.MobileScrollLock` component so other modal-on-mobile UI like the default and publish options sidebars can declare scroll lock as well. `Popover.MobileScrollLock` is used in the editor layout to declare scroll lock when sidebars are displayed on mobile.

Scroll locking is accomplished by adding a `lockscroll` class to the `<html>` and `<body>` elements. Unlocking restores the pre-lock scroll position.

This solution is body-specific, so scroll bleed to other ancestor elements is still possible. We can prevent scroll bleed on mobile by preventing the default action on `touchmove` events, but I'd like to learn more first to be sure we don't break other mobile interactions involving `touchmove`.

## Additional comments

* I'd prefer to encapsulate what makes a mobile screen in `Popover.MobileScrollLock` but did not because that knowledge is currently duplicated between editor modules and `components/popover`.
* Scroll bleed only appears to be an issue in portrait orientation because `overflow: hidden` is applied to the body at a min-width of 600px.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

### CODE REVIEW

**Comment by jasmussen** [comment]
Nice work, thanks for working on this! 

> 🚨 This is currently a non-starter because setting overflow: hidden on the body resets the scroll position to zero. It's surprising when you're scrolled halfway down a post, open and close the block inserter, and find you're now scrolled to the top of the post. I am considering a solution for this. 

Is this actually the case, though? I know the following is just testing in the inspector, but for me it works fine to set `overflow: hidden;` on the `body` element:

https://cloudup.com/cYbErRWbcOC

In my experience, the only thing that resets the scroll position is if for whatever reason the main scrolling canvas changes _height_. Is it when it lands on a real device that it stops working?

**Comment by brandonpayton** [comment]
Hi @jasmussen, I'm sorry for the delay in response. Here is what I'm seeing when I the `lockscroll` class to the `body` and `html` elements.
https://cloudup.com/cb-hZYMuoaw

The scroll unfortunately resets.

I am thinking saving and restoring the body's `scrollTop` is a reasonable workaround, but it might interfere with other logic that does things like scroll elements into view. An example of that is the [`url-input`](https://github.com/WordPress/gutenberg/blob/master/blocks/url-input/index.js#L167) block. 

One idea is to lock and unlock scroll in `componentWillUpdate` since things like scroll-into-view likely need to wait for `componentDidUpdate`. This is getting too clever IMO, but I don't currently have a cleaner idea for working around this issue.

What do you think?

**Comment by jasmussen** [comment]
It seems I was mistaken indeed. On Android and in the Chrome simulator, applying `overflow: hidden;` _only_ to the `body` element achieves the desired effect, however you're right, it has to be applied also to the `html` element for this to work on iOS. And incidentally the latter is what resets the scroll position. That's unfortunate.

It does seem like, perhaps, your approach is what we'll have to do — store the scroll position and then restore it on modal-close. This one seemed promising: https://benfrain.com/preventing-body-scroll-for-modals-in-ios/

**Comment by brandonpayton** [comment]
Regarding your link, I love the idea of preventing the default action for `touchmove`, especially because we'd only need to worry about touches that start within the modal and we wouldn't have to meddle with the scroll.

I've wondered whether it would be too clever though. Here are a couple of questions I need to answer:

[ ] Are `touchmove` events emitted when dragging an element? If so, we could interfere with drag and drop.
[ ] Is it possible to scroll diagonally on mobile? If so, we can interfere with horizontal scrolling if we prevent default on a diagonal scroll because it would cause vertical scroll bleed.

In the meantime, I'm thinking it probably makes sense to save and restore body scroll position.

**Comment by jasmussen** [comment]
Thanks for keeping on this Brandon.

Just to clarify — _you_ should decide, based on your knowledge of this, what is the best programmatic approach to solving this problem. While I might sometimes suggest something to explore, it's merely a casual suggestion, and being mostly a designer I'd always defer to you and others on the code implementation. So go forth with what you feel will work best 🏅

**Comment by brandonpayton** [comment]
@jasmussen I think this is ready for consideration.

I'm not 100% comfortable with saving and restoring scroll position because I'm afraid it will interfere with other components that scroll-into-view, but in my testing, scrolling after inserting a block is working as I'd expect. But if we're able to land #5316, I think I can updated `Popover.MobileScrollLock` similarly and restore scroll at an earlier point in the React lifecycle to reduce the likelihood of this issue.

**Comment by brandonpayton** [comment]
@jasmussen I take that back. I had previously tested on an Android device but am seeing issues restoring scroll position on the latest. It seems to restore but not actually update the display until I interact with something. I'll reply again once I know more.

**Comment by brandonpayton** [comment]
I committed a fix. I thought this was working previously but don't see how. The scroll position needed to be restored on `document.scrollingElement`. It appears to be the same as `document.body` on iOS, at least in the editor, but it is the same as `document.documentElement` on Android.

I noticed I don't have a test for restoring scroll position, so I'm adding one.

**Comment by brandonpayton** [comment]
I am not sure how to reasonably test restoring scroll position, but I may be able to add an e2e test that resizes the window to a mobile viewport width to trigger scroll locking.

It might be worth landing this as-is though.

**Comment by jasmussen** [comment]
In my testing, this is absolutely working, and working really really well. Because it's so good, I strongly agree it would be good to get in asap. Adding a quick code review but experience wise this seems great. Thank you!

File: components/popover/mobile-scroll-lock.js
Comment (gziolo) [inline components/popover/mobile-scroll-lock.js]: There is a folder with all Higher-Order Components: https://github.com/WordPress/gutenberg/tree/master/components/higher-order. We also try to come up with a pattern for naming such helpers. It seems like in this case `withScrollLock` would be a good pick. In general, we prefix all HOCs with `with` in the majority of cases. The only exception is when a component might not be rendered when a given condition occurs. In such case `if` is the preferred prefix. @aduth might even update our docs somewhere, but I'm sure he commented about it at least twice.

Comment (gziolo) [reply components/popover/mobile-scroll-lock.js]: Okey, I now see that it is an independent component. Skip my previous comment in this context :)

File: edit-post/components/layout/index.js
Comment (gziolo) [inline edit-post/components/layout/index.js]: There is also `ifViewportMatch` HOC which might be used to wrap this component. In that case, it would never mount when the viewport is larger than small (mobile). It leaves `layoutHasOpenSidebar` part in here. My question is if we can convert this component into HOC that would be applied to the popover instead? In that case we would naturally combine two related concepts. This `MobileScrollLock` would be a sibling component that renders only on mobile and when any popover is actually rendered.

File: components/popover/index.js
Comment (gziolo) [inline components/popover/index.js]: If we want to make it reusable for other use cases it should be exposed as its own component to avoid confusion. I was sure it is specific to the Popover component.

File: edit-post/components/layout/index.js
Comment (brandonpayton) [reply edit-post/components/layout/index.js]: I would love to use `ifViewportMatch` here, but my understanding is that it depends on resolving a circular dependency between `@wordpress/components` and `@wordpress/viewport` (see #5316). We can probably get most of the way there though:

1. I can update `MobileScrollLock` to be enabled if present.
2. Make users of `MobileScrollLock` responsible for only adding for mobile viewports. I believe we only use it in two places right now.

Once we can use `ifViewportMatch`, we can push that requirement into `MobileScrollLock`.

File: components/popover/index.js
Comment (brandonpayton) [reply components/popover/index.js]: I think you're right and will make the change. The original thinking was that it was ideally only for popover but could be used for Popover-like components.

**Comment by brandonpayton** [comment]
Thanks for the review, @gziolo!

**Comment by brandonpayton** [comment]
This has been updated in response to review comments.

File: components/mobile-scroll-lock/README.md
Comment (gziolo) [inline components/mobile-scroll-lock/README.md]: `<Popover.MobileScrollLock />` wasn't update to reflect the latest changes.

File: components/mobile-scroll-lock/README.md
Comment (gziolo) [inline components/mobile-scroll-lock/README.md]: I still think that using HOC would be the cleanest solution because the scroll lock behavior would be applied on top of the component that needs it.

```js
function Sidebar() {
	return (
		<div>
			Sidebar Content!
		</div>
	);
}
export default withMobileScrollLock( Sidebar );
```

File: components/popover/index.js
Comment (gziolo) [inline components/popover/index.js]: This component would benefit from using this new HOC `withViewportMatch`, too.

It's still tricky because it depends on another prop. I think we could add additional checks to HOC:
```js
export default withMobileScrollLock(
    ( props ) => Boolean( props.expandOnMobile ),
)( Popover );
```

File: edit-post/components/layout/index.js
Comment (gziolo) [inline edit-post/components/layout/index.js]: When using HOC `withMobileScrollLock` we wouldn't have to put this logic in here at all, which is quite disconnected from the sidebar component that needs them.

File: components/mobile-scroll-lock/README.md
Comment (gziolo) [inline components/mobile-scroll-lock/README.md]: I think that it can be renamed to `ScrollLock` as it is no longer dependent on any mobile related logic :)

File: components/mobile-scroll-lock/index.js
Comment (gziolo) [inline components/mobile-scroll-lock/index.js]: Yes, it's now a general purpose scroll lock component :)

**Comment by gziolo** [comment]
Thanks adding your changes to the PR 👍 

> I would love to use `ifViewportMatch` here, but my understanding is that it depends on resolving a circular dependency between `@wordpress/components` and `@wordpress/viewport` (see #5316). 

Yes, this needs to be resolved separately first. Then we would be able to refactor what we have in this PR and introduce `withMobileScrollLock` HOC, which could look like:

```js
const withMobileScrollLock = ( predicate) => ( WrappedComponent ) => { 
    const MobileScrollLock = compose(
        ifViewportMatches( '< small' ),
        ifCondition( predicate ),
    )( ScrollLock );
    const EnhancedComponent = ( props ) => (
        <Fragment>
            <MobileScrollLock { ...props } />
            <WrappedComponent { ...props } />
        </Fragment>
    );

    EnhancedComponent.displayName = getWrapperDisplayName( WrappedComponent, 'withMobileScrollLock' );

    return EnhancedComponent;
};
```

I don't want to block this PR until we find a viable solution for `ifViewportMatch`. You can skip temporarily my comments related to `withMobileScrollLock` HOC. Let's change the name `MobileScrollLock` to `ScrollLock` and update all corresponding docs and we should be good to go 💯

**Review by gziolo — CHANGES_REQUESTED** [review]
See above.

**Comment by brandonpayton** [comment]
Hi @gziolo, I believe I've address all your review comments.

I like your idea of a `withMobileScrollLock` HOC. Thanks for taking time to talk about it and work through how it would be applied to various `<ScrollLock>` consumers.

File: components/scroll-lock/test/index.js
Comment (gziolo) [inline components/scroll-lock/test/index.js]: There is no risk, all test suites are isolated when you use Jest :)

File: edit-post/components/layout/index.js
Comment (gziolo) [inline edit-post/components/layout/index.js]: Nit: can we extract inline variable for those 3 flags? We use it twice.

**Review by gziolo — APPROVED** [review]
This looks great after a few iterations. Thanks for addressing all feedback 👍 
I noticed that @jasmussen has tested it already, so I'll assume it work as advertised :)

File: components/scroll-lock/test/index.js
Comment (brandonpayton) [reply components/scroll-lock/test/index.js]: I was very excited by this, but in my testing, it seems like the global `document` persists between tests and suites.

File: edit-post/components/layout/index.js
Comment (brandonpayton) [reply edit-post/components/layout/index.js]: Thanks for catching that. I pushed a fix.

File: components/scroll-lock/test/index.js
Comment (gziolo) [reply components/scroll-lock/test/index.js]: Interesting, I will check it once in `master`, out of curiosity :)

**Comment by mtias** [comment]
Let's get this for 2.5 to make sure there are no unintended issues.

**Comment by afercia** [comment]
> Scroll locking is accomplished by adding a lockscroll class to the <html> and <body> elements. 

Just to note that core uses a CSS class `modal-open` for this, and only on the body.

**Comment by brandonpayton** [comment]
Thanks, that's good to know, @afercia. In my testing, it was necessary to add this to the document element as well to effectively prevent scroll bleed on iOS devices. It's possible there is something I missed. Any thoughts here?

File: components/scroll-lock/test/index.js
Comment (brandonpayton) [reply components/scroll-lock/test/index.js]: I added a class to the body in one test, checked for its presence in `beforeEach`, and found the class on the body in the second test. Are you up for letting me know if you find different behavior? I'd love to take advantage of isolated environment.

**Comment by brandonpayton** [comment]
I need to rebase this and relocate the rules to component CSS. I'm planning to do that this morning. Then this should be ready to go.

**Pull request merged**
